openupgrade_scripts/scripts/sale_purchase/16.0.1.0/upgrade_analysis_work.txt

```diff
@@ -0,0 +1,9 @@
+---Models in module 'sale_purchase'---
+---Fields in module 'sale_purchase'---
+sale_purchase / product.template         / service_to_purchase (boolean) : not stored anymore
+# DONE: post-migration, convert to company-dependent field
+---XML records in module 'sale_purchase'---
+DEL ir.model.constraint: sale_purchase.constraint_product_template_service_to_purchase
+# DONE: post-migration, delete XML-ID
```

Comment:
What this actually means is that the DB constraint has been removed.

v15: https://github.com/odoo/odoo/blob/06919a232251cbd68a28500757aef5cd2e3363a0/addons/sale_purchase/models/product_template.py#L13
v16: https://github.com/odoo/odoo/blob/17151f5c369dce7ca5d38c5fd0db8dceab5cfab7/addons/sale_purchase/models/product_template.py#L14

so you need to remove the constraint as well in the DB.